### ISSUE TITLE
`utilities.data.extract_batch_size` can silently fail -- maybe it needs to be less automatic?

### ISSUE THREAD

**Issue by ronif** [PyTorchLightning/pytorch-lightning#10093]
## Motivation

Hey! So from what I understand `extract_batch_size` is used by `ResultCollection` to reduce/re-weight various logging metrics such as validation loss, etc.

We had an issue where our batch collation function aggregated Tensors into a list instead of stacking them (needed because of size mismatches), and `extract_batch_size` would recursively crawl through the batch object until it found a Tensor and returned its first dimention size -- which is wrong (the list length is correct batch size). So we would silently get "random" batch sizes and the val_loss logging becomes incorrect and then the LR scheduling / early stopping incorrectly triggers etc.

### Pitch

1. Maybe there should be some warning / note in the documentation (didn't find any)
2. If the batch object isn't a tensor (like a dict/dataclass), maybe try to `getattr(batch, 'batch_size')` and only then produce a warning and try to infer it automatically.
